### ISSUE TITLE
Unable to cross-compile for aarch64

### ISSUE THREAD

**Issue by garyscetbon** [google/mediapipe#926]
I've installed Ubuntu 20.04 on a Raspberry 4 and I'm trying to cross-compile a Mediapipe app that uses OpenCV from a Docker container running on Ubuntu 18.04.

I've copied the opencv_*.so files from `/usr/lib/aarch64-linux-gnu/` on the Raspberry to `/usr/lib/aarch64-linux-gnu/' in the container. 

My `third_party/opencv_linux.BUILD` looks like:
```
cc_library(
    name = "opencv",
    srcs = glob(
        [
            "lib/aarch64-linux-gnu/libopencv_core.so",
            "lib/aarch64-linux-gnu/libopencv_calib3d.so",
            "lib/aarch64-linux-gnu/libopencv_features2d.so",
            "lib/aarch64-linux-gnu/libopencv_highgui.so",
            "lib/aarch64-linux-gnu/libopencv_imgcodecs.so",
            "lib/aarch64-linux-gnu/libopencv_imgproc.so",
            "lib/aarch64-linux-gnu/libopencv_video.so",
            "lib/aarch64-linux-gnu/libopencv_videoio.so",
        ],
    ),
    hdrs = glob(["include/opencv2/**/*.h*"]),
    includes = ["include/"],
    linkstatic = 1,
    visibility = ["//visibility:public"],
)
```

When I run: 
```
bazel build -c opt --crosstool top=@crosstool//:toolchains --compiler=gcc --cpu=aarch64 --define MEDIAPIPE_DISABLE_GPU=1 --define 'libunwind=true' mediapipe/examples/desktop/hand_tracking:hand_tracking_cpu 
```

I get the error: 
```
INFO: Analyzed target //mediapipe/examples/desktop/hand_tracking:hand_tracking_cpu (1 packages loaded, 59 targets configured).
INFO: Found 1 target...
ERROR: /mediapipe/mediapipe/examples/desktop/hand_tracking/BUILD:27:1: Linking of rule '//mediapipe/examples/desktop/hand_tracking:hand_tracking_cpu' failed (Exit 1) aarch64-linux-gnu-gcc failed: error executing command /usr/bin/aarch64-linux-gnu-gcc @bazel-out/aarch64-opt/bin/mediapipe/examples/desktop/hand_tracking/hand_tracking_cpu-2.params

Use --sandbox_debug to see verbose messages from the sandbox
bazel-out/aarch64-opt/bin/mediapipe/examples/desktop/libdemo_run_graph_main.a(demo_run_graph_main.o):demo_run_graph_main.cc:function cv::Mat::~Mat(): error: undefined reference to 'cv::fastFree(void*)'
```

Has anyone any idea on how to make it work?

**Comment by mcclanahoochie** [comment]
Have you take a look at the [coral](https://github.com/google/mediapipe/tree/master/mediapipe/examples/coral) build instructions (seems likely)?   Most of that process would apply here.

Do you have the exact same opencv version installed in both places?

Does compiling just the hello_world work?

It seems to be a linker issue, so please double check that the libs are copied correctly (sometimes symlinks cause issues).

The raspberry pi 4 should be powerful enough (i.e. enough RAM) to compile directly on the device, if that is an option for you.. (the main issue is bazel needs lots of RAM. it's been done on coral 1GB devices too, by specially allocating HD space as extra ram).

**Comment by garyscetbon** [comment]
Thanks for your help @mcclanahoochie I ended up spotting the error: I was adding the `new_local_repository` in the wrong file! Maybe I skimmed the docs [here](https://google.github.io/mediapipe/getting_started/install.html#installing-on-debian-and-ubuntu) a bit too fast :) Once I moved that rule to the right place (`mediapipe/WORKSPACE`, not `mediapipe/third_party/opencv_linux.BUILD`) and fixed a compiler issue re the zlib directory (issue [#1406](https://github.com/tensorflow/serving/issues/1406)), it all worked out fine. 

The only problem I'm facing now is that there's a 3-4s lag and the fps is very low when I compile the `desktop/hand_tracking` example for my Raspberry Pi 4 (running on Ubuntu 20.04). I'm wondering if I'm using the edge tpu capabilities correctly. 

Here's my compile command: 
```
bazel build -c opt --crosstool_top=@crosstool//:toolchains --compiler=gcc --cpu=aarch64 --define MEDIAPIPE_DISABLE_GPU=1 --copt -DMEDIAPIPE_EDGE_TPU --copt=-flax-vector-conversions mediapipe/examples/desktop/hand_tracking:hand_tracking_cpu
```

Any advice on how to increase computing speed with the Coral USB key for that hand tracking example?

**Comment by mcclanahoochie** [comment]
Glad you got things running.

I don't think the hand_tracking model is fully supported by coral. Also, the coral TPU requires the model to be converted to 8bit and then recompiled from that into a new edge_tpu file; I guess you have done that?

Double check the logs, because I doubt the usb stick is being used, and if it is, only a few ops are actually running on the accelerator. Unless there has been an update to the coral edge tpu compiler (which is possible), we found there are some ops used in the hand tracking/detection models that were not supported by edge TPU, then delegating those ops (an all following) to the CPU. 

So, my suspicion is that the model is just running 100% on the cpu if you haven't recompiled for edge tpu, and if you have, then probably there are unsupported ops preventing running fully on TPU.

**Comment by garyscetbon** [comment]
The face and object detection examples work perfectly and way faster than the hand tracking one but your suspicions are correct: I haven't converted the tflite to an 8bit one. I'll keep you posted if anything arides when I try so! Still pretty impressed by the relative speed you can get when running the current tflite model on a simple raspberry CPU though. Thanks for the support!

**Comment by zwfcrazy** [comment]
@garyscetbon @mcclanahoochie

**Comment by zwfcrazy** [comment]
@garyscetbon @mcclanahoochie 
Hey guys, I am trying to cross compile mediapipe for raspberrypi 4 with Raspbian 32-bit (armhf), but I keep getting linking errors.
I installed opencv libs with armhf on my host machine (ubuntu 18.04 amd64) and modified opencv_linux.BUILD (lib/arm-linux-gnueabihf). By doing so I managed to link libopencv files, however, the linker could not find lib files required by opencv. So I had to add linkopts to opencv_linux.BUILD. Eventually I got
`linkopts = [
        "-Wl,-rpath-link=/usr/lib/arm-linux-gnueabihf",
        "-Wl,-rpath-link=/usr/lib",
        "-Wl,-rpath-link=/lib/arm-linux-gnueabihf",
        ],`
This solved most of the link errors, but finally I got std link errors which do not know how to sovle.
